File: backend/auth_service/server/auth_service.go
package server

import (
	"context"
	"errors"
	"fmt"
	"io"
	"log"
	"os"

	pb "auth_service/generated"
	"auth_service/models"
	"auth_service/repository"
	"auth_service/services"
	"auth_service/utils"

	"github.com/go-redis/redis"
)

// AuthServiceServer dengan Redis Client sebagai Dependency Injection
type AuthServiceServer struct {
	pb.UnimplementedAuthServiceServer
	RedisClient    *redis.Client // Tambahkan Redis sebagai field
	sekolahService services.SekolahService
	authService    services.AuthService
	userProfile    services.UserProfileService
}

// Constructor untuk AuthServiceServer dengan Redis
func NewAuthServiceServer(redisClient *redis.Client) *AuthServiceServer {
	return &AuthServiceServer{RedisClient: redisClient}
}

type SchoolRegistration struct {
	SchoolName string `json:"school_name"`
	AdminEmail string `json:"admin_email"`
}

func (s *AuthServiceServer) Login(ctx context.Context, req *pb.LoginRequest) (*pb.LoginResponse, error) {
	username := req.GetUsername()
	password := req.Password

	resp, err := s.authService.Login(username, password)
	if err != nil {
		log.Printf("Error username/password salah : %v", err)
		return nil, err
	}
	// Generate JWT
	token, err := utils.GenerateJWT(resp)
	if err != nil {
		return nil, errors.New("failed to generate token")
	}
	return &pb.LoginResponse{
		Token: token,
		Ok:    true,
		User: &pb.User{
			UserId:    resp.ID,
			Username:  resp.Username,
			Email:     resp.Email,
			Role:      resp.Role,
			SekolahId: resp.SekolahID,
		},
	}, nil
}

func (s *AuthServiceServer) Register(ctx context.Context, req *pb.RegisterRequest) (*pb.RegisterResponse, error) {
	// Ambil data dari request
	user := req.GetUser()
	sekolah := req.GetSekolah()
	var query = repository.SekolahQuery{
		Npsn: sekolah.Npsn,
	}
	// Cek apakah sekolah sudah ada
	sekolahModel, err := s.sekolahService.GetSekolah(query)
	if err != nil {
		if errors.Is(err, services.ErrNotFound) {
			if user.Role == "admin" {
				// Buat sekolah baru
				sekolahModel, err = s.sekolahService.CreateSekolah(&models.Sekolah{
					NPSN:            sekolah.Npsn,
					NamaSekolah:     sekolah.NamaSekolah,
					SekolahIDEnkrip: sekolah.SekolahIdEnkrip,
					Kecamatan:       sekolah.Kecamatan,
					Kabupaten:       sekolah.Kabupaten,
					Propinsi:        sekolah.Propinsi,
					KodeKecamatan:   sekolah.KodeKecamatan,
					KodeKab:         sekolah.KodeKab,
					KodeProp:        sekolah.KodeProp,
					AlamatJalan:     sekolah.AlamatJalan,
					Status:          sekolah.Status,
				})
				if err != nil {
					log.Printf("Gagal membuat sekolah: %v", err)
					return nil, fmt.Errorf("gagal membuat sekolah: %w", err)
				}
			} else {
				// Pendaftaran siswa
				return nil, fmt.Errorf("sekolah belum terdaftar di aplikasi")
			}
		} else {
			log.Printf("Server error saat mencari sekolah: %v", err)
			return nil, fmt.Errorf("server error: %w", err)
		}
	}
	// Hubungkan user dengan sekolah
	userModel := &models.User{
		Username:  user.Username,
		Email:     user.Email,
		Role:      user.Role,
		SekolahID: int32(sekolahModel.ID),
		Password: user.Password,
	}

	// Cek jika role user adalah admin dan apakah sudah ada admin
	if userModel.Role == "admin" {
		adminExists, err := s.authService.IsAdminExists(sekolahModel.ID)
		if err != nil {
			log.Printf("Error mengecek admin: %v", err)
			return nil, fmt.Errorf("server error: %w", err)
		}
		if adminExists {
			return nil, fmt.Errorf("admin sudah ada untuk sekolah ini")
		}

		// Registrasi admin
		if err := s.authService.RegisterAdmin(userModel); err != nil {
			log.Printf("Error registrasi admin: %v", err)
			return nil, fmt.Errorf("gagal registrasi admin: %w", err)
		}
	} else if userModel.Role == "siswa" {
		// Registrasi siswa
		if err := s.authService.Register(userModel); err != nil {
			log.Printf("Error registrasi siswa: %v", err)
			return nil, fmt.Errorf("gagal registrasi siswa: %w", err)
		}
	}

	// Hubungkan user dengan profil
	userProfileModel := &models.UserProfile{
		UserID: userModel.ID,
	}

	if err := s.userProfile.CreateUserProfile(userProfileModel); err != nil {
		log.Printf("Error membuat user profile: %v", err)
		return nil, fmt.Errorf("server error saat membuat user profile")
	}

	var response *pb.RegisterResponse
	if userModel.Role == "admin" {
		token, err := utils.GenerateJWT(userModel)
		if err != nil {
			return nil, errors.New("failed to generate token")
		}
		response = &pb.RegisterResponse{
			Token: token,
			Ok:    true,
			User: &pb.User{
				UserId:    userModel.ID,
				Username:  userModel.Username,
				Email:     userModel.Email,
				Role:      userModel.Role,
				SekolahId: userModel.SekolahID,
			},
		}
	} else {
		response = &pb.RegisterResponse{
			Ok: true,
		}
	}
	log.Println("User berhasil didaftarkan")

	// // Siapkan data registrasi sekolah untuk dikirim ke Redis
	// registration := SchoolRegistration{
	// 	SchoolName: sekolah.GetNamaSekolah(), // Menggunakan metode GetNamaSekolah()
	// 	AdminEmail: user.GetEmail(),          // Menggunakan metode GetEmail()
	// }

	// // Konversi data ke JSON
	// data, err := json.Marshal(registration)
	// if err != nil {
	// 	log.Printf("Error marshalling registration data: %v", err)
	// 	return nil, fmt.Errorf("gagal memproses data registrasi")
	// }

	// // Kirim data ke Redis
	// err = s.RedisClient.Publish("school_registration", data).Err()
	// if err != nil {
	// 	log.Printf("Error publishing to Redis: %v", err)
	// 	return nil, fmt.Errorf("gagal mengirim data ke sistem antrian")
	// }

	// fmt.Println("Registration message published successfully!")

	return response, nil
}

// GetUserProfile - Mengambil profil pengguna berdasarkan UserID
func (s *AuthServiceServer) GetUserPofile(ctx context.Context, req *pb.GetUserProfileRequest) (*pb.GetUserProfileResponse, error) {
	profile, err := s.userProfile.GetUserProfileByID(req.UserId)
	if err != nil {
		log.Printf("Error fetching user profile: %v", err)
		return nil, errors.New("failed to retrieve user profile")
	}
	return &pb.GetUserProfileResponse{
		UserProfile: &pb.UserProfile{
			UserId:   profile.UserID,
			Nama:     profile.Nama,
			Jk:       profile.JK,
			Phone:    profile.Phone,
			TptLahir: profile.TptLahir,
			// TglLahir: profile.TglLahir.Format(),
			AlamatJalan: profile.AlamatJalan,
			KotaKab:     profile.KotaKab,
			Prov:        profile.Prov,
			KodePos:     profile.KodePos,
			NamaAyah:    profile.NamaAyah,
			NamaIbu:     profile.NamaIbu,
		},
	}, nil
}

// UpdateUserProfile - Memperbarui profil pengguna berdasarkan UserID
func (s *AuthServiceServer) UpdateUserPofile(ctx context.Context, req *pb.UpdateUserProfileRequest) (*pb.UpdateUserProfileResponse, error) {
	// Debugging: Cek nilai request yang diterima
	log.Printf("Received UpdateUserProfile request: %+v\n", req)

	// Cek apakah req atau req.UserProfile kosong
	if req == nil {
		log.Println("Request is nil")
		return nil, errors.New("invalid request: request is nil")
	}

	if req.UserProfile == nil {
		log.Println("UserProfile is nil in request")
		return nil, errors.New("invalid request: user profile is nil")
	}

	profile, err := s.userProfile.GetUserProfileByID(int64(req.GetUserId()))
	if err != nil {
		log.Printf("Error fetching user profile: %v", err)
		return nil, errors.New("user profile not found")
	}
	// profi := req.GetUserProfile()
	// log.Println(profi)
	// Perbarui data profil berdasarkan input
	profile.Nama = req.UserProfile.Nama
	profile.JK = req.UserProfile.Jk
	profile.Phone = req.UserProfile.Phone
	profile.TptLahir = req.UserProfile.TptLahir
	profile.AlamatJalan = req.UserProfile.AlamatJalan
	profile.KotaKab = req.UserProfile.KotaKab
	profile.Prov = req.UserProfile.Prov
	profile.KodePos = req.UserProfile.KodePos
	profile.NamaAyah = req.UserProfile.NamaAyah
	profile.NamaIbu = req.UserProfile.NamaIbu

	// Simpan perubahan ke database
	err = s.userProfile.UpdateUserProfile(profile)
	if err != nil {
		log.Printf("Error updating user profile: %v", err)
		return nil, errors.New("failed to update user profile")
	}

	return &pb.UpdateUserProfileResponse{
		Message: "Updated",
	}, nil
}

// UploadUserPhotoProfile - Mengunggah foto profil pengguna
func (s *AuthServiceServer) UploadUserPhotoProfile(stream pb.AuthService_UploadUserPhotoProfileServer) error {
	var userID int32
	var filePath string
	var file *os.File
	// var err error

	for {
		req, err := stream.Recv()
		if err == io.EOF {
			// Jika selesai menerima file, kirim respons sukses
			return stream.SendAndClose(&pb.UploadUserPhotoResponse{
				Status:  "OK",
				FileUrl: filePath,
			})
		}
		if err != nil {
			log.Printf("Error receiving file chunk: %v", err)
			return err
		}

		// Ambil User ID dari request pertama
		if userID != 0 {
			userID = req.UserId
			filePath = fmt.Sprintf("uploads/%d_profile.jpg", userID)
			// Simpan ke database
			curentProfile, err := s.userProfile.GetUserProfileByID(int64(userID))
			if err != nil {
				log.Printf("Error getting user profile: %v", err)
			}
			err = s.userProfile.UpdateUserProfile(&models.UserProfile{
				ID:             curentProfile.ID,
				ProfilePicture: filePath,
			})
			if err != nil {
				log.Printf("Error updating user profile: %v", err)
			}
			// Buat file baru
			file, err = os.Create(filePath)
			if err != nil {
				log.Printf("Error creating file: %v", err)
				return err
			}
			defer file.Close()
		}

		// Tulis chunk ke file
		_, err = file.Write(req.FileChunk)
		if err != nil {
			log.Printf("Error writing file: %v", err)
			return err
		}
	}
}

func (s *AuthServiceServer) GetSekolah(ctx context.Context, req *pb.GetSekolahRequest) (*pb.GetSekolahResponse, error) {
	var query = repository.SekolahQuery{
		Npsn:      req.GetNpsn(),
		SekolahID: int(req.GetSekolahId()),
	}

	sekolah, err := s.sekolahService.GetSekolah(query)
	if err != nil {
		log.Printf("Error fetching school data: %v", err)
		return nil, errors.New("failed to retrieve school data")
	}

	return &pb.GetSekolahResponse{
		Nama: sekolah.NamaSekolah,
		SekolahData: &pb.Sekolah{
			SekolahId:       int32(sekolah.ID),
			SekolahIdEnkrip: sekolah.SekolahIDEnkrip,
			Kecamatan:       sekolah.Kecamatan,
			Kabupaten:       sekolah.Kabupaten,
			Propinsi:        sekolah.Propinsi,
			KodeKecamatan:   sekolah.Kecamatan,
			AlamatJalan:     sekolah.AlamatJalan,
			KodeKab:         sekolah.KodeKab,
			KodeProp:        sekolah.KodeProp,
			NamaSekolah:     sekolah.NamaSekolah,
			Status:          sekolah.Status,
			Npsn:            sekolah.NPSN,
		},
	}, nil
}

File: frontend/src/store/authService.js
import axios from "axios";
// const baseURL = "http://localhost:8080/api/v1";
const api = axios.create({
  baseURL: "http://localhost:8080/api/v1", // Pastikan menggunakan protokol HTTPS
  withCredentials: true, // Untuk mengirim cookie atau credensial
  headers: {
    "Content-Type": "application/json",
    "Content-Type": "Authorization",
  },
});

const state = {
  token: localStorage.getItem("token") || null,
  userRole: localStorage.getItem("userRole") || null,
  user: localStorage.getItem("user")
    ? JSON.parse(localStorage.getItem("user"))
    : null,
  userProfile: JSON.parse(localStorage.getItem("userProfile")) || null, // Ambil dari localStorage
  sekolah: JSON.parse(localStorage.getItem("sekolah")) || null, // Ambil dari localStorage
  refreshToken: null,
  loading: false,
  error: null,
};

const mutations = {
  setUser(state, user) {
    state.user = user;
    localStorage.setItem("user", JSON.stringify(user)); // Simpan user ke localStorage
  },
  setUserRole(state, userRole) {
    state.userRole = userRole;
    localStorage.setItem("userRole", userRole);
  },
  setUserProfile(state, userProfile) {
    state.userProfile = userProfile;
    localStorage.setItem("userProfile", JSON.stringify(userProfile));
  },

  setToken(state, token) {
    state.token = token;
    localStorage.setItem("token", token);
  },
  setRefreshToken(state, refreshToken) {
    state.refreshToken = refreshToken;
  },
  clearAuthData(state) {
    state.token = null;
    state.refreshToken = null;
    state.user = null;
    localStorage.removeItem("token"); // Hapus token saat logout
    localStorage.removeItem("userRole"); // Hapus userRole saat logout
    localStorage.removeItem("userProfile"); // Hapus userProfile saat logout
    localStorage.removeItem("user");
    localStorage.removeItem("sekolah");
    localStorage.removeItem("tabelTenant");
  },
  SET_SEKOLAH(state, sekolah) {
    state.sekolah = sekolah;
    localStorage.setItem("sekolah", JSON.stringify(sekolah));
  },
  // setSekolah(state, sekolah) {
  //   state.sekolah = sekolah;
  //   localStorage.setItem("sekolah", JSON.stringify(sekolah));
  // },
  SET_LOADING(state, isLoading) {
    state.loading = isLoading;
  },
  SET_ERROR(state, error) {
    state.error = error;
  },
};

const actions = {
  async login({ commit }, credentials) {
    try {
      commit("SET_LOADING", true);
      commit("SET_ERROR", null);
      const response = await api.post("/auth/login", credentials);
      if (response.data.ok) {
        commit("setToken", response.data.token);
        commit("setUser", response.data.user);
        commit("setUserRole", response.data.user.role);
        return true;
      }
    } catch (error) {
      commit("SET_ERROR", error.response?.data || "Terjadi kesalahan");
      console.error("Login failed:", error.response ? error.response.data : error.message);
      return error; // Menghindari mengembalikan error langsung
      // return false; // Menghindari mengembalikan error langsung
    } finally {
      commit("SET_LOADING", false);
    }
  },
  async logout({ commit }) {
    try {
      console.log("clear auth data");
      // await axios.post("/auth/logout"); // Opsional: panggil endpoint logout server
      commit("clearAuthData");
    } finally {
      commit("clearAuthData");
    }
  },
  async registerAdmin({ commit }, payload) {
    try {
      const response = await api.post("/auth/register", payload);
      // console.log(response);

      if (response.data.ok) {
        commit("setToken", response.data.token);
        // Simpan informasi pengguna setelah login
        commit("setUser", response.data.user);
        commit("setUserRole", response.data.user.role);
      }
      return response.data;
    } catch (error) {
      console.error("Registration failed:", error.response.data);
      throw error.response.data;
    }
  },
  // Fitur baru ceknpsn
  async ceknpsn({ commit }, npsn) {
    commit("SET_LOADING", true);
    commit("SET_ERROR", null);
    try {
      const response = await api.get(`/sekolah`, {
        params: {
          npsn: npsn,
        },
      });
      commit("SET_SEKOLAH", response.data);
      return response.data; // Mengembalikan data sekolah
    } catch (error) {
      commit("SET_ERROR", error.response?.data || "Terjadi kesalahan");
      console.error("Gagal mengambil data NPSN:", error);
      return null;
    } finally {
      commit("SET_LOADING", false);
    }
  },
  async getSekolahByID({ commit }, sekolahId) {
    commit("SET_LOADING", true);
    commit("SET_ERROR", null);
    try {
      const response = await api.get(`/sekolah`, {
        params: {
          sekolah_id: sekolahId,
        },
      });
      // console.log(response);
      commit("SET_SEKOLAH", response.data);
      // return response.data; // Mengembalikan data sekolah
    } catch (error) {
      commit("SET_ERROR", error.response?.data || "Terjadi kesalahan");
      console.error("Gagal mengambil data ID Sekola:", error);
      return null;
    } finally {
      commit("SET_LOADING", false);
    }
  },
  //  Ambil Profil Pengguna
  async getUserProfile({ commit }, userID) {
    try {
      const response = await api.get(`/user/${userID}/profile`); // Pastikan endpoint benar
      commit("setUserProfile", response.data.userProfile);
      return response.data;
    } catch (error) {
      console.error("Gagal mengambil profil pengguna:", error);
      return null;
    }
  },

  // Update Profil Pengguna
  async updateUserProfile({ commit }, updatedProfile) {
    console.log("Mengirim data ke server:", updatedProfile);
    try {
      const response = await api.put(`/user/${updatedProfile.userId}/profile`, {
        user_id: updatedProfile.userId, // Sesuai dengan .proto
        user_profile: {
          // Harus dikirim dalam objek "user_profile"
          nama: updatedProfile.nama,
          jk: updatedProfile.jk,
          phone: updatedProfile.phone,
          tpt_lahir: updatedProfile.tptLahir,
          alamat_jalan: updatedProfile.alamatJalan,
          kota_kab: updatedProfile.kotaKab,
          prov: updatedProfile.prov,
          kode_pos: updatedProfile.kodePos,
          nama_ayah: updatedProfile.namaAyah,
          nama_ibu: updatedProfile.namaIbu,
        },
      });

      console.log("Response dari server:", response.data);

      if (response.data.status === "success") {
        commit("setUserProfile", response.data.user_profile);
        return response.data;
      } else {
        console.error("Gagal memperbarui profil:", response.data);
        return null;
      }
    } catch (error) {
      console.error(
        "Gagal memperbarui profil pengguna:",
        error.response?.data || error.message
      );
      return null;
    }
  },
  // Upload Foto Profil Pengguna
  async uploadUserPhotoProfile({ commit }, file) {
    try {
      const formData = new FormData();
      formData.append("photo", file);

      const response = await api.post("/user/profile/photo", formData, {
        headers: {
          "Content-Type": "multipart/form-data",
        },
      });

      commit("setUser", response.data);
      return response.data;
    } catch (error) {
      console.error("Gagal mengunggah foto profil:", error);
      return null;
    }
  },
};

const getters = {
  isAuthenticated(state) {
    return !!state.token;
  },
  userRole(state) {
    return state.userRole;
  },
  getSekolah(state) {
    return state.sekolah;
  },
  getUserProfile(state) {
    const userData = { ...state.user, ...state.userProfile };
    return userData;
  },
  isLoading: (state) => state.loading,
  getError: (state) => state.error,
};

export default {
  namespaced: true,
  state,
  mutations,
  actions,
  getters,
};
